### ISSUE TITLE
Merge finite_source_gradient branch

### ISSUE THREAD

**Issue by rpoleski** [rpoleski/MulensModel#89]
There are tasks to be done before we can merge [finite_source_gradient branch](https://github.com/rpoleski/MulensModel/compare/finite_source_gradient)

- [x] remove `sandbox/`
- [x] should new functions be public or private?
- [x] review changes in `fitdata.py` - assigned to @rpoleski
- [x] FSPLDerivs class - change name
- [x] FSPLDerivs class - add docstring
- [x] FSPLDerivs - docstrings for all private and public functions
- [x] add docstings for all added functions, including `class TestFSPLGradient`
- [x] `_get_magnification_curve()` in FSPLDerivs - remove calls to private properties of Model class (probably by adding public functions/properties to Model)
- [x] review added tests - assigned to @rpoleski 
- [x] _Not doing this_: remove calls to `FSPLDerivs._get_B0_prime()` and `FSPLDerivs._get_B1_prime()` from `class TestFSPLGradient`
- [x] check coverage of added code - see next post
- [x] make sure tests pass (currently 2 tests fail in `tests/test_Model.py`: `test_BLPS_02()` and `test_model_binary_and_finite_sources()`; The former is corrected in the master branch, so no need to worry about it here.)
- [x] make sure that gradient for FSPL model with `t_star` instead of `rho` fails
- [x] remove duplicated code: `FitData._read_B0B1_file()` and `PointLens._read_B0B1_file()` and extra interpolation_table.dat files - assigned to @rpoleski
- [x] review recent commits: https://github.com/rpoleski/MulensModel/commit/522b22106c6a764192c069312b404ce6bfcdd76f https://github.com/rpoleski/MulensModel/commit/dca11ecca20bec3d5d157f8654b7dc4bceac506e https://github.com/rpoleski/MulensModel/commit/ccf649cda360a1bb2c42df4546c2b6fec650a7ba - assigned to @rpoleski
- [x] comment in line 461 of `source/MulensModel/tests/test_Model.py` - should it be removed? @jenniferyee 
- [x] `FSPL_Derivatives.__init__()` - move actual content to the top of this block? - No (see comment from Aug 20).
- [x] `fitdata.py` - remove calls to private functions of other classes: `self.model._parallax` (simple), `self._magnification_curve._methods_for_epochs()`
- [x] `Model.set_magnification_methods_parameters()` - comments 1. and 2.
- [x] remove warnings raised by unit tests (15 of them on 2023.08.11)
- [x] run flake8
- [x] At the end: version, tests, readme.md, sphinx.

Then we can merge :)

**Comment by rpoleski** [comment]
Coverage now misses:

 - [ ] 1st if in `FitData.get_d_A_d_params_for_point_lens_model`
 - [ ] if in `FSPLDerivs._get_dataset_satellite_skycoord`
 - [ ] a few in `FSPLDerivs._get_b0_gamma_b1_and_derivs`
 - [ ] `FSPLDerivs.get_gradient`
 - [ ] `FitData.get_d_A_d_u_for_FSPL_model`
 - [ ] in `FitData.get_d_A_d_u_for_point_lens_model`
 - [ ] `FitData.data_magnification`
 - [ ] `LimbDarkeningCoeffs.__repr__`
 
Remaining to be checked: `model.py` and `pointlens.py`.

**Comment by jenniferyee** [comment]
Note about duplicated code:  FitData._read_B0B1_file() and PointLens._read_B0B1_file()

Path forward: Ideally, figure out where this code was originally, then figure out how to call that function. But it might be an architecture issue, in which case, it needs to be punted to Version 3.

**Comment by rpoleski** [comment]
`PointLens._read_B0B1_file()` was the original one. We can move it to a separate class that will be called by both `FitData` and `PointLens`.

**Comment by rpoleski** [comment]
@jenniferyee Why there are functions defined _inside_ `FSPLDervis.__init__()`? In my opinion is makes code messier than defining them in `FSPLDervis`. They also use `self` even though it's not their argument.

**Comment by jenniferyee** [comment]
There are function inside of FSPLDerivs.__init__() because that is the only place they are used. It's a convention for organizing all of the initialization functions. Otherwise, you get a lot of functions in the class with no general purpose, so you have to figure out why they are there. By nesting them inside __init__() it is clear that they are used in __init__(). Also, they automatically know about "self" because __init__() knows about self.

**Comment by jenniferyee** [comment]
Changed `FSPLDerivs` to `_FSPLDerivs` because the user shouldn't need to see this class.
@rpoleski Does that satisfy your request to "change the name"?

**Comment by jenniferyee** [comment]
@rpoleski I don't think we should remove calls to private functions/properties in model (e.g. Lines 728--731). Why? Because this is all related to the underlying architecture issue I raise previously. We could make all these things public, but we'd just have to change them back later when we refactor the code. This was always the problem with this branch: it is a lot of work to fix all of these little things, but it would be better to invest that work in the underlying problem instead of hacking a fix now.

**Comment by jenniferyee** [comment]
@rpoleski  I have some questions.
> 
>     * [ ]  should new functions be public or private?

I'm in favor of making things private unless we know they need to be public.


>     * [ ]  FSPLDerivs class - change name

What's wrong with it?

> 
>     * [ ]  `_get_magnification_curve()` in FSPLDerivs - remove calls to private properties of Model class (probably by adding public functions/properties to Model)

I don't think we should remove calls to private functions/properties in model (e.g. Lines 728--731). Why? Because this is all related to the underlying architecture issue I raise previously. We could make all these things public, but we'd just have to change them back later when we refactor the code. This was always the problem with this branch: it is a lot of work to fix all of these little things, but it would be better to invest that work in the underlying problem instead of hacking a fix now.

> 
>     * [ ]  `FitData.get_d_A_d_u_for_PSPL_model()` - revert function name change (or change version to v3.0.0)

@rpoleski Do you mean PSPL --> point_lens? because `FitData.get_d_A_d_u_for_point_lens_model` already (still) exists. 


> 
>     * [ ]  remove calls to `FSPLDerivs._get_B0_prime()` and `FSPLDerivs._get_B1_prime()` from `class TestFSPLGradient`

Why? I thought more tests are better?

> 
>     * [ ]  remove duplicated code: `FitData._read_B0B1_file()` and `PointLens._read_B0B1_file()`

1. `FitData._read_B0B1_file()` reads a different file than `PointLens._read_B0B1_file()`

2. @rpoleski set up the architecture of this. I'm not sure how to fix it.

>     * [ ]  docstrings for added functions

Duplicate action item?

**Comment by rpoleski** [comment]
- Functions inside `FSPLDervis.__init__()`. I don't see a problem with having a few functions in the class that are called only by `__init__()`. If they have proper names (e.g., `_setup_...`, `_set_starting...` etc.) then it's clear what their purpose is. The disadvantage of having them inside `__init__()` is that the parameters are not directly stated. I may copy-paste it to some other code and this can cause problems.
- `FSPLDerivs` class - we agreed to have meaningful function, class, and variable names. I'm not sure I'll easily understand that name in a few months.
- Removing calls to private functions of other classes. The private functions can be changed at any point in any way. If one class calls such methods of the other class, then it's easy to produce side-effects when changing one class. The idea of making all parameters of magnification methods of `Model` class public was around for some time, so in this case I don't see a problem with adding such a function. I can do it myself.
- `FitData._read_B0B1_file()` vs. `PointLens._read_B0B1_file()` - both read  `os.path.join(mm.DATA_PATH, 'interpolation_table_b0b1_v3.dat')`.
- Remove duplicated code: `FitData._read_B0B1_file()` and `PointLens._read_B0B1_file()` - I can do that. Task assigned. 

To be continued (3 tasks remaining).

**Comment by rpoleski** [comment]
Issue to be discussed: how to make public these 3 `Model` properties: `._methods`, `.default_magnification_method`, and `._methods_parameters`?

There are two possible approaches:

1. `@property` decorator and making current `set_*`  functions deprecated,
2. adding `get_methods_settings()` function that returns all 3 values.

The issue I see with the first approach is that `set_magnification_methods()` has `source` parameter, which allows to set methods for just one of the sources in a binary source event.

**Comment by jenniferyee** [comment]
> Issue to be discussed: how to make public these 3 `Model` properties: `._methods`, `.default_magnification_method`, and `._methods_parameters`?
> 
> There are two possible approaches:
> 
> 1. `@property` decorator and making current `set_*`  functions deprecated,
> 2. adding `get_methods_settings()` function that returns all 3 values.
> 
> The issue I see with the first approach is that `set_magnification_methods()` has `source` parameter, which allows to set methods for just one of the sources in a binary source event.

Made get_magnification_methods_parameters() and get_magnification_methods() that parallel the behavior of the set_() functions, as well as properties that return the _ variables. See commit [5d62d60](https://github.com/rpoleski/MulensModel/commit/5d62d6082b1092998e6424d3a941ec4c6fcd44f3)

**Comment by jenniferyee** [comment]
@rpoleski I think I fixed the Warnings raised by the unit tests, but you should check.

**Comment by rpoleski** [comment]
> `FSPL_Derivatives.__init__()` - move actual content to the top of this block?

No, because the functions have to be defined before they're used. In my opinion, this part of the code is less clear than it would be in a standard approach (i.e., no functions defined in `__init__()`).

**Comment by rpoleski** [comment]
@jenniferyee added comments in `Model.set_magnification_methods_parameters()` some time ago. Answers to the problems marked there:

1. Witt & Mao methods were omitted by mistake
2. There is a list of all existing methods, not the methods for which parameters can be set. The latter is limited in `MagnificationCurve`.

**Comment by rpoleski** [comment]
@jenniferyee you have added both `Model.get_magnification_methods_parameters()` and `Model.methods_parameters`. Both of them do basically the same thing - give public access to `Model._methods_parameters`. **Is there a need to have 2 such methods?**
This is the last question to be answered or resolved. After that we can merge (then run sphinx and make new release). If you want to see the new code added, go to [pointlensfinitesource.py](https://github.com/rpoleski/MulensModel/blob/finite_source_gradient/source/MulensModel/pointlensfinitesource.py).

**Comment by jenniferyee** [comment]
There was already `Model.set_magnification_methods_parameters()` , which leads to the conclusion that there should be a "get".

However, when I originally copy-pasted the code for creating a MagnificationCurve into FSPLDerivs, it called `set_magnification_methods_parameters(model._methods_parameters). So, this needed to be able to directly access `_methods_parameters` without calling a private function, so I created `Model.methods_parameters`. Later, I realized this call was irrelevant for the point lens magnification methods, and the call no longer exists. 

Conclusion: `Model.methods_parameters` can be deleted (but be sure to also remove the related unit tests).
